File: packages/sdk/src/administrator/payments.ts
import { QelosSDKOptions } from '../types';
import BaseSDK from '../base-sdk';
import {
  IPlan, ISubscription, IInvoice, ICoupon,
  BillableEntityType, SubscriptionStatus, InvoiceStatus, BillingCycle,
} from '@qelos/global-types';

export interface AdminCheckoutRequest {
  planId: string;
  billingCycle: BillingCycle;
  /** When omitted, the authenticated user / default billable entity is used (same as public checkout). */
  billableEntityType?: BillableEntityType;
  billableEntityId?: string;
  couponCode?: string;
  successUrl?: string;
  cancelUrl?: string;
  /**
   * Admin-only. For plans with `dynamic: true`, sets the `dynamicAmount` on the
   * pending subscription before initiating checkout. Regular users cannot set this;
   * use `setSubscriptionDynamicAmount()` + a separate `checkout()` call instead.
   */
  amount?: number;
}

export interface CreateSubscriptionData {
  planId: string;
  billingCycle: BillingCycle;
  billableEntityType: BillableEntityType;
  billableEntityId: string;
  /** Required for plans with `dynamic: true` before checkout can be initiated. */
  dynamicAmount?: number;
  couponCode?: string;
}

export interface AdminCheckoutResponse {
  subscriptionId: string;
  checkoutUrl?: string;
  clientToken?: string;
}

export default class QlPaymentsAdmin extends BaseSDK {
  constructor(private options: QelosSDKOptions) {
    super(options);
  }

  // --- Plans ---

  getPlans(query?: { isActive?: boolean }) {
    const qs = query ? `?${new URLSearchParams(query as any)}` : '';
    return this.callJsonApi<IPlan[]>(`/api/plans${qs}`);
  }

  getPlan(planId: string) {
    return this.callJsonApi<IPlan>(`/api/plans/${planId}`);
  }

  createPlan(data: Omit<IPlan, '_id' | 'tenant' | 'created'>) {
    return this.callJsonApi<IPlan>('/api/plans', {
      method: 'post',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(data),
    });
  }

  updatePlan(planId: string, data: Partial<IPlan>) {
    return this.callJsonApi<IPlan>(`/api/plans/${planId}`, {
      method: 'put',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(data),
    });
  }

  deletePlan(planId: string) {
    return this.callJsonApi<IPlan>(`/api/plans/${planId}`, { method: 'delete' });
  }

  /**
   * Start subscription checkout (same as `sdk.payments.checkout`, plus optional billable-entity overrides for admins).
   * For plans with `dynamic: true`, `amount` is required.
   */
  checkout(params: AdminCheckoutRequest) {
    return this.callJsonApi<AdminCheckoutResponse>('/api/checkout', {
      method: 'post',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(params),
    });
  }

  // --- Subscriptions ---

  getSubscriptions(query?: {
    billableEntityType?: BillableEntityType;
    billableEntityId?: string;
    status?: SubscriptionStatus;
  }) {
    const qs = query ? `?${new URLSearchParams(query as any)}` : '';
    return this.callJsonApi<ISubscription[]>(`/api/subscriptions${qs}`);
  }

  getSubscription(subscriptionId: string) {
    return this.callJsonApi<ISubscription>(`/api/subscriptions/${subscriptionId}`);
  }

  /**
   * Creates a pending subscription on behalf of any billable entity. Admins can
   * also set `dynamicAmount` here for dynamic plans, making the subscription
   * immediately ready for checkout.
   */
  createSubscription(data: CreateSubscriptionData) {
    return this.callJsonApi<ISubscription>('/api/subscriptions', {
      method: 'post',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(data),
    });
  }

  cancelSubscription(subscriptionId: string) {
    return this.callJsonApi<ISubscription>(
      `/api/subscriptions/${subscriptionId}/cancel`,
      { method: 'put' },
    );
  }

  /**
   * Sets or updates the dynamic amount for a pending subscription. Only admins
   * can call this. Must be done before a user can complete checkout on a dynamic plan.
   */
  setSubscriptionDynamicAmount(subscriptionId: string, amount: number) {
    return this.callJsonApi<ISubscription>(
      `/api/subscriptions/${subscriptionId}/dynamic-amount`,
      {
        method: 'put',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ amount }),
      },
    );
  }

  // --- Invoices ---

  getInvoices(query?: {
    billableEntityType?: BillableEntityType;
    billableEntityId?: string;
    status?: InvoiceStatus;
  }) {
    const qs = query ? `?${new URLSearchParams(query as any)}` : '';
    return this.callJsonApi<IInvoice[]>(`/api/invoices${qs}`);
  }

  getInvoice(invoiceId: string) {
    return this.callJsonApi<IInvoice>(`/api/invoices/${invoiceId}`);
  }

  // --- Coupons ---

  getCoupons(query?: { isActive?: boolean }) {
    const qs = query ? `?${new URLSearchParams(query as any)}` : '';
    return this.callJsonApi<ICoupon[]>(`/api/coupons${qs}`);
  }

  getCoupon(couponId: string) {
    return this.callJsonApi<ICoupon>(`/api/coupons/${couponId}`);
  }

  createCoupon(data: Omit<ICoupon, '_id' | 'tenant' | 'created' | 'currentRedemptions'>) {
    return this.callJsonApi<ICoupon>('/api/coupons', {
      method: 'post',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(data),
    });
  }

  updateCoupon(couponId: string, data: Partial<ICoupon>) {
    return this.callJsonApi<ICoupon>(`/api/coupons/${couponId}`, {
      method: 'put',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(data),
    });
  }

  deleteCoupon(couponId: string) {
    return this.callJsonApi<ICoupon>(`/api/coupons/${couponId}`, { method: 'delete' });
  }
}

File: packages/python-sdk/qelos_sdk/administrator/payments.py
from __future__ import annotations

import json
from typing import Any, Dict, List, Optional
from urllib.parse import urlencode

from ..base_sdk import BaseSDK
from ..types import QelosSDKOptions


class QlPaymentsAdmin(BaseSDK):
    """Admin payment management: plans, subscriptions, invoices, and coupons."""

    def __init__(self, options: QelosSDKOptions) -> None:
        super().__init__(options)

    # --- Plans ---

    async def get_plans(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List all plans."""
        qs = "?" + urlencode(query) if query else ""
        return await self.call_json_api(f"/api/plans{qs}")

    async def get_plan(self, plan_id: str) -> Dict[str, Any]:
        """Get a plan by ID."""
        return await self.call_json_api(f"/api/plans/{plan_id}")

    async def create_plan(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new plan."""
        return await self.call_json_api(
            "/api/plans",
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps(data),
        )

    async def update_plan(self, plan_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """Update a plan."""
        return await self.call_json_api(
            f"/api/plans/{plan_id}",
            method="PUT",
            headers={"content-type": "application/json"},
            body=json.dumps(data),
        )

    async def delete_plan(self, plan_id: str) -> Dict[str, Any]:
        """Delete a plan."""
        return await self.call_json_api(f"/api/plans/{plan_id}", method="DELETE")

    # --- Subscriptions ---

    async def get_subscriptions(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List subscriptions with optional filters."""
        qs = "?" + urlencode(query) if query else ""
        return await self.call_json_api(f"/api/subscriptions{qs}")

    async def get_subscription(self, subscription_id: str) -> Dict[str, Any]:
        """Get a subscription by ID."""
        return await self.call_json_api(f"/api/subscriptions/{subscription_id}")

    async def cancel_subscription(self, subscription_id: str) -> Dict[str, Any]:
        """Cancel a subscription."""
        return await self.call_json_api(
            f"/api/subscriptions/{subscription_id}/cancel", method="PUT"
        )

    # --- Invoices ---

    async def get_invoices(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List invoices with optional filters."""
        qs = "?" + urlencode(query) if query else ""
        return await self.call_json_api(f"/api/invoices{qs}")

    async def get_invoice(self, invoice_id: str) -> Dict[str, Any]:
        """Get an invoice by ID."""
        return await self.call_json_api(f"/api/invoices/{invoice_id}")

    # --- Coupons ---

    async def get_coupons(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List coupons."""
        qs = "?" + urlencode(query) if query else ""
        return await self.call_json_api(f"/api/coupons{qs}")

    async def get_coupon(self, coupon_id: str) -> Dict[str, Any]:
        """Get a coupon by ID."""
        return await self.call_json_api(f"/api/coupons/{coupon_id}")

    async def create_coupon(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new coupon."""
        return await self.call_json_api(
            "/api/coupons",
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps(data),
        )

    async def update_coupon(self, coupon_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """Update a coupon."""
        return await self.call_json_api(
            f"/api/coupons/{coupon_id}",
            method="PUT",
            headers={"content-type": "application/json"},
            body=json.dumps(data),
        )

    async def delete_coupon(self, coupon_id: str) -> Dict[str, Any]:
        """Delete a coupon."""
        return await self.call_json_api(f"/api/coupons/{coupon_id}", method="DELETE")
